Xamarin.PropertyEditing/ViewModels/PanelViewModel.cs

```diff
@@ -15,5 +11,177 @@ public PanelViewModel (IEditorProvider provider)
+				props = props.Where (MatchesFilter);
+
+			props = props.OrderBy (vm => vm.Property.Name);
+
+			this.arranged.Clear ();
+			foreach (var grouping in props.GroupBy (GetGroup).OrderBy (g => g.Key)) {
+				this.arranged.Add (grouping);
+			}
+
+			ArrangedPropertiesChanged?.Invoke (this, EventArgs.Empty);
+		}
+
+		protected override void OnRemoveProperties (IEnumerable<PropertyViewModel> properties)
+		{
+			foreach (PropertyViewModel vm in properties) {
+				string g = GetGroup (vm);
+				var grouping = this.arranged[g] as ObservableGrouping<string, PropertyViewModel>;
+				if (grouping != null) {
+					this.arranged.Remove (g, vm);
+				}
+			}
+
+			ArrangedPropertiesChanged?.Invoke (this, EventArgs.Empty);
+		}
+
+		protected override void OnClearProperties ()
+		{
+			this.arranged.Clear ();
+
+			ArrangedPropertiesChanged?.Invoke (this, EventArgs.Empty);
+		}
+
+		private readonly Dictionary<PropertyArrangeMode, HashSet<string>> expandedGroups = new Dictionary<PropertyArrangeMode, HashSet<string>> ();
+		private readonly ObservableLookup<string, PropertyViewModel> arranged = new ObservableLookup<string, PropertyViewModel> {
+			ReuseGroups = true
+		};
+
+		private PropertyArrangeMode arrangeMode;
+		private string filterText;
+
+		private void Arrange()
+		{
+			this.arranged.Clear ();
+
+			OnAddProperties (Properties);
+		}
+
+		private enum FilterState
+		{
+			Unknown,
+			Shorter,
+			Longer
+		}
+
+		private void Filter (string oldFilter)
+		{
+			FilterState state = FilterState.Unknown;
+			if (String.IsNullOrWhiteSpace (oldFilter) || FilterText.StartsWith (oldFilter))
+				state = FilterState.Longer;
+			else if (oldFilter.StartsWith (FilterText))
+				state = FilterState.Shorter;
+
+			if (state != FilterState.Shorter) {
+				var toRemove = new List<PropertyViewModel> ();
+				foreach (var g in this.arranged) {
+					foreach (var vm in g) {
+						if (!MatchesFilter (vm))
+							toRemove.Add (vm);
+					}
+				}
+
+				OnRemoveProperties (toRemove);
+			}
+
+			if (state != FilterState.Longer) {
+				OnAddProperties (Properties);
+			}
+		}
+
+		private string GetGroup (PropertyViewModel vm)
+		{
+			return (ArrangeMode == PropertyArrangeMode.Name) ? "0" : vm.Category;
+		}
+
+		private ObservableGrouping<string, PropertyViewModel> GetGroupList (PropertyViewModel propertyVm)
+		{
+			switch (ArrangeMode) {
+				case PropertyArrangeMode.Name:
+					return (arranged.Count > 0) ? arranged[0] as ObservableGrouping<string, PropertyViewModel> : null;
+				case PropertyArrangeMode.Category:
+					return arranged[propertyVm.Category] as ObservableGrouping<string, PropertyViewModel>;
+
+				default:
+					return null;
+			}
+		}
+
+		private bool MatchesFilter (PropertyViewModel vm)
+		{
+			if (String.IsNullOrWhiteSpace (FilterText))
+				return true;
+
+			string filter = FilterText.ToLower ();
+			if (ArrangeMode == PropertyArrangeMode.Category && vm.Category != null && vm.Category.ToLower().Contains (filter)) {
```

Comment:
We should review all `ToLower`, `ToUpper` calls, and also be careful about anything that deals with culture specific number formatting.